### PR TITLE
Support for extension

### PR DESCRIPTION
**Pull request by thmarx**
Hey,

I've added simple support for custom extension and want to know what you think about it. 
Current state is really simple and maybe there are some code changes needed. 

As an example extension, I've used the Notification extension from a pull request (https://github.com/gitbucket/markedj/pull/12).

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -50,107 +50,85 @@
...
+		<groupId>org.assertj</groupId>
+		<artifactId>assertj-core</artifactId>
+		<version>3.24.2</version>
+		<scope>test</scope>
+	</dependency>
   </dependencies>
 
-  <profiles>
-    <profile>
-      <id>release-sign-artifacts</id>
```
Comment (takezoe) [inline pom.xml:57]: Why would you remove the profile in this pull request?

**Comment by takezoe** [comment]
Sorry, I missed this pull request for long 🙇‍♂️ This is very interesting so I'm now taking a close look to see if we can merge this for the next markedj release.

File: pom.xml (hunk 1 above)
Comment (thmarx) [reply pom.xml]: sorry, this happened accidentally during the merge. I just undo the modification.

File: src/main/java/io/github/gitbucket/markedj/extension/Extension.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ *
+ * @author t.marx
+ */
+public interface Extension {
+	
+	public String getName();
+	
+	public Map<String, Rule> enhanceRules(Map<String, Rule> existingRules);
```
Comment (takezoe) [inline src/main/java/io/github/gitbucket/markedj/extension/Extension.java:32]: Is adding rules actually necessary although `Extension` is called directly in `Parser` and `Lexer`?

File: src/main/java/io/github/gitbucket/markedj/extension/Extension.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ *
+ * @author t.marx
+ */
+public interface Extension {
+	
+	public String getName();
```
Comment (takezoe) [inline src/main/java/io/github/gitbucket/markedj/extension/Extension.java:30]: This method isn't used?

File: src/main/java/io/github/gitbucket/markedj/extension/Extension.java (hunk 2 above)
Comment (thmarx) [reply src/main/java/io/github/gitbucket/markedj/extension/Extension.java]: You are right, this is not needed anymore. code is removed.

File: src/main/java/io/github/gitbucket/markedj/Lexer.java [hunk 4]
```diff
@@ -170,6 +171,12 @@ protected void token(String src, boolean top, boolean bq, LexerContext context){
                 }
             }
 
+			{
+				for (Extension extension : options.getExtensions()) {
+					src = extension.lex(src, context, this::token);
```
Comment (takezoe) [inline src/main/java/io/github/gitbucket/markedj/Lexer.java:176]: Should continue if any rule is applied?

File: src/main/java/io/github/gitbucket/markedj/Parser.java [hunk 5]
```diff
@@ -1,23 +1,25 @@
...
     protected Renderer renderer;
+    protected Map<String, Rule> rules;
 
     public Parser(Options options, Renderer renderer){
         this.options = options;
         this.renderer = renderer;
     }
 
     public String parse(Stack<Token> src, Map<String, Lexer.Link> links){
-        Map<String, Rule> rules;
```
Comment (takezoe) [inline src/main/java/io/github/gitbucket/markedj/Parser.java:20]: No need to move `rules` to fields. Or we can move setting up rules to the constructor if we move `rules` to fields.

**Comment by takezoe** [comment]
@thmarx Sorry, I have two more comments.

File: src/main/java/io/github/gitbucket/markedj/Lexer.java (hunk 4 above)
Comment (thmarx) [reply src/main/java/io/github/gitbucket/markedj/Lexer.java]: The interface needed to be expanded here, but I think it should fit now.

File: src/main/java/io/github/gitbucket/markedj/Lexer.java [hunk 6]
```diff
@@ -170,6 +171,18 @@ protected void token(String src, boolean top, boolean bq, LexerContext context){
                 }
             }
 
+			{
+				boolean doContinue = false;
+				for (Extension extension : options.getExtensions()) {
+					Extension.LexResult result = extension.lex(src, context, this::token);
+					src = result.getSource();
+					doContinue = doContinue || result.matches();
```
Comment (takezoe) [inline src/main/java/io/github/gitbucket/markedj/Lexer.java:179]: Thanks. I thought just comparing the returned `src` with the original one would be okay but I like this interface as it's more clear.

By the way, I also think it's okay to continue immediately if `result.matches()` is true. Are there cases that multiple extensions are applied at the same time?

Comment (thmarx) [reply src/main/java/io/github/gitbucket/markedj/Lexer.java:179]: You're absolutely right, there should be no valid cases to check more extensions after the first match.

**Review by takezoe — APPROVED** [review]

**Comment by takezoe** [comment]
Thank you for your contribution!

**Comment by takezoe** [comment]
By the way, it would be great if you can also update README since this added a new option and the notification extension.

**Pull request merged**